Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,616 @@
+  /// If the path has no root and only includes a single component,
+  /// returns an empty FilePath.
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/usr/bin/ls => /usr/bin`
+  ///   * `/foo        => /`
+  ///   * `/           => /`
+  ///   * `foo         => ""`
+  /// * Windows:
+  ///   * `C:\foo\bar.exe                 => C:\foo`
+  ///   * `C:\                            => C:\`
+  ///   * `\\server\share\folder\file.txt => \\server\share\folder`
+  ///   * `\\server\share\                => \\server\share\`
+  public __consuming func removingLastComponent() -> FilePath {
+    var copy = self
+    copy.removeLastComponent()
+    return copy
+  }
+
+  /// In-place mutating variant of `removingLastComponent`.
+  ///
+  /// If `self` only contains a root, does nothing and returns `false`.
+  /// Otherwise removes `lastComponent` and returns `true`.
+  ///
+  /// Example:
+  ///
+  ///     var path = "/usr/bin"
+  ///     path.removeLastComponent() == true  // path is "/usr"
+  ///     path.removeLastComponent() == true  // path is "/"
+  ///     path.removeLastComponent() == false // path is "/"
+  ///
+  @discardableResult
+  public mutating func removeLastComponent() -> Bool {
+    defer { _invariantCheck() }
+    guard let lastRel = lastComponent else { return false }
+    _storage.removeSubrange(lastRel._slice.indices)
+    _removeTrailingSeparator()
+    return true
+  }
+}
+
+extension FilePath.Component {
+  /// The extension of this file or directory component.
+  ///
+  /// If `self` does not contain a `.` anywhere, or only
+  /// at the start, returns `nil`. Otherwise, returns everything after the dot.
+  ///
+  /// Examples:
+  ///   * `foo.txt    => txt`
+  ///   * `foo.tar.gz => gz`
+  ///   * `Foo.app    => app`
+  ///   * `.hidden    => nil`
+  ///   * `..         => nil`
+  ///
+  public var `extension`: String? {
+    guard let range = _extensionRange() else { return nil }
+    return _slice[range].string
+  }
+
+  /// The non-extension portion of this file or directory  component.
+  ///
+  /// Examples:
+  ///   * `foo.txt => foo`
+  ///   * `foo.tar.gz => foo.tar`
+  ///   * `Foo.app => Foo`
+  ///   * `.hidden => .hidden`
+  ///   * `..      => ..`
+  ///
+  public var stem: String {
+    _slice[_stemRange()].string
+  }
+}
+
+extension FilePath {
+
+  /// The extension of the file or directory last component.
+  ///
+  /// If `lastComponent` is `nil` or one of the special path components
+  /// `.` or `..`, `get` returns `nil` and `set` does nothing.
+  ///
+  /// If `lastComponent` does not contain a `.` anywhere, or only
+  /// at the start, `get` returns `nil` and `set` will append a
+  /// `.` and `newValue` to `lastComponent`.
+  ///
+  /// Otherwise `get` returns everything after the last `.` and `set` will
+  /// replace the extension.
+  ///
+  /// Examples:
+  ///   * `/tmp/foo.txt                 => txt`
+  ///   * `/Appliations/Foo.app/        => app`
+  ///   * `/Appliations/Foo.app/bar.txt => txt`
```

Comment:
```suggestion
  ///   * `/Applications/Foo.app/        => app`
  ///   * `/Applications/Foo.app/bar.txt => txt`
```

Reply:
You'll need to re-indent the rest of this comment to get the => arrows to line up